prototype/mobile_endpoint/case/cleanup.py

```diff
@@ -0,0 +1,48 @@
+import logging
+from mobile_endpoint.case import const
+from mobile_endpoint.case.xml.parser import KNOWN_PROPERTIES
+
+logger = logging.getLogger(__name__)
+
+
+def primary_actions(case):
+    return filter(lambda a: a.action_type != const.CASE_ACTION_REBUILD,
+                  case.actions)
+
+
+def reset_state(case):
+    """
+    Clear known case properties, and all dynamic properties
+    """
+    dynamic_properties = set([k for action in case.actions for k in action.updated_unknown_properties.keys()])
+    for k in dynamic_properties:
+        try:
+            delattr(case, k)
+        except KeyError:
+            pass
+        except AttributeError:
+            # 'case_id' is not a valid property so don't worry about spamming
+            # this error.
+            if k != 'case_id':
+                logger.error(
+                    "Cannot delete attribute '%(attribute)s' from case '%(case_id)s'" % {
+                        'case_id': case.id,
+                        'attribute': k,
+                    }
+                )
+
+    # already deleted means it was explicitly set to "deleted",
+    # as opposed to getting set to that because it has no actions
+    already_deleted = case.doc_type == 'CommCareCase-Deleted' and primary_actions(case)
```

Comment:
we should think about how we'll deal with deletions in the new models. moving to a new table would make sense to me though.